packages/apollo/src/commands/service/check.ts

```diff
@@ -253,112 +260,198 @@ export default class ServiceCheck extends ProjectCommand {
+                // If we're `federated`, then run composition validation. When we're using composition
+                // validation we'll receive a schema has that represents the composed schema.
+                if (ctx.federationSchemaHash) {
+                  schemaCheckSchemaVariables = {
+                    schemaHash: ctx.federationSchemaHash
+                  };
+                } else {
+                  // This is _not_ a `federated` schema. Resolve the schema given `config.tag`.
+                  task.output = "Resolving schema";
+                  schema = await project.resolveSchema({ tag: config.tag });
+                  if (!schema) {
+                    throw new Error("Failed to resolve schema");
+                  }
 
-                  task.output = "Validating schema";
-                  const variables: CheckSchemaVariables = {
-                    id: configName,
-                    // @ts-ignore
-                    // XXX Looks like TS should be generating ReadonlyArrays instead
-                    schema: introspectionFromSchema(schema).__schema,
-                    tag: flags.tag,
-                    gitContext: await gitInfo(this.log),
-                    frontend: flags.frontend || config.engine.frontend,
-                    ...(historicParameters && { historicParameters })
+                  schemaCheckSchemaVariables = {
+                    schema: introspectionFromSchema(schema)
+                      .__schema as IntrospectionSchemaInput
                   };
-                  const { schema: _, ...restVariables } = variables;
-                  this.debug("Variables sent to Engine:");
-                  this.debug(restVariables);
+                }
+
+                await gitInfo(this.log);
+
+                const historicParameters = validateHistoricParams({
+                  validationPeriod: flags.validationPeriod,
+                  queryCountThreshold: flags.queryCountThreshold,
+                  queryCountThresholdPercentage:
+                    flags.queryCountThresholdPercentage
+                });
+
+                task.output = "Validating schema";
+
+                const variables: CheckSchemaVariables = {
+                  id: configName,
+                  tag: flags.tag,
+                  gitContext: await gitInfo(this.log),
+                  frontend: flags.frontend || config.engine.frontend,
+                  ...(historicParameters && { historicParameters }),
+                  // `variables` will either contain `schemaHash` or `schema`, not both.
+                  ...schemaCheckSchemaVariables
+                };
+
+                const { schema: _, ...restVariables } = variables;
+                this.debug("Variables sent to Engine:");
+                this.debug(restVariables);
+                if (schema) {
                   this.debug("SDL of introspection sent to Engine:");
                   this.debug(printSchema(schema));
+                }
```

Comment:
It would be helpful to log the schemaHash here.

Since it's debug mode, I'd also think that the command to introspect/download the full schema from manual inspection would complete the experience

Reply:
I added the schema has; I don't know if we have a command to actually use it.